src/serializer/ebml.rs

```diff
@@ -15,381 +20,245 @@ pub fn vint_size(i: u64) -> u8 {
     id: u64,
-    num: i64,
-) -> Result<(&mut [u8], usize), GenError> {
-    do_gen!(
-        input,
-        gen_call!(gen_vid, id) >> gen_call!(gen_vint, 8) >> gen_be_i64!(num)
-    )
+    expected_size: u8,
+    f: G,
+) -> impl Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError>
+where
+    G: Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError>,
+{
+    move |input| {
+        let (buf, ofs_len) = gen_vid(id)(input)?;
+        let (buf, start) = gen_skip(expected_size as usize)((buf, ofs_len))?;
+        let (buf, end) = f((buf, start))?;
+        gen_at_offset(ofs_len, gen_ebml_size(expected_size, end - start))((buf, end))
+    }
 }
 
-pub fn gen_i32(
-    input: (&mut [u8], usize),
+pub(crate) fn gen_ebml_uint_l(
     id: u64,
-    num: i32,
-) -> Result<(&mut [u8], usize), GenError> {
-    do_gen!(
-        input,
-        gen_call!(gen_vid, id) >> gen_call!(gen_vint, 4) >> gen_be_i32!(num)
-    )
+    num: u64,
+    expected_size: u8,
+) -> impl Fn((&mut [u8], usize)) -> Result<(&mut [u8], usize), GenError> {
+    move |input| {
+        let needed_bytes = vint_size(expected_size as u64);
+
+        let (buf, ofs_len) = gen_vid(id)(input)?;
+        let (buf, start) = gen_skip(needed_bytes as usize)((buf, ofs_len))?;
+        let (buf, end) = gen_uint(num)((buf, start))?;
+        gen_at_offset(ofs_len, gen_ebml_size(expected_size, end - start))((buf, end))
+    }
 }
 
-pub fn gen_i16(
-    input: (&mut [u8], usize),
+pub(crate) fn gen_ebml_uint(
     id: u64,
-    num: i16,
-) -> Result<(&mut [u8], usize), GenError> {
-    do_gen!(
-        input,
-        gen_call!(gen_vid, id) >> gen_call!(gen_vint, 2) >> gen_be_i16!(num)
-    )
+    num: u64,
+) -> impl Fn((&mut [u8], usize)) -> Result<(&mut [u8], usize), GenError> {
+    gen_ebml_uint_l(id, num, vint_size(num))
 }
 
-pub fn gen_i8(input: (&mut [u8], usize), id: u64, num: i8) -> Result<(&mut [u8], usize), GenError> {
-    do_gen!(
-        input,
-        gen_call!(gen_vid, id) >> gen_call!(gen_vint, 1) >> gen_be_i8!(num)
-    )
+pub(crate) fn gen_ebml_int(
+    id: u64,
+    num: i64,
+) -> impl Fn((&mut [u8], usize)) -> Result<(&mut [u8], usize), GenError> {
+    move |input| {
+        let expected_size = vint_size(num as u64);
+        let needed_bytes = vint_size(expected_size as u64);
+
+        let (buf, ofs_len) = gen_vid(id)(input)?;
+        let (buf, start) = gen_skip(needed_bytes as usize)((buf, ofs_len))?;
+        let (buf, end) = gen_int(num)((buf, start))?;
+        gen_at_offset(ofs_len, gen_ebml_size(expected_size, end - start))((buf, end))
+    }
 }
 
-pub fn gen_f64(
-    input: (&mut [u8], usize),
+pub(crate) fn gen_ebml_str<'a>(
     id: u64,
-    num: f64,
-) -> Result<(&mut [u8], usize), GenError> {
-    do_gen!(
-        input,
-        gen_call!(gen_vid, id) >> gen_call!(gen_vint, 8) >> gen_be_f64!(num)
-    )
+    s: &'a str,
+) -> impl Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
```

Comment:
```suggestion
    s: &'b str,
) -> impl Fn((&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> + 'b {
```

or something like that maybe?